drafts/draft-ietf-ippm-ioam-data-integrity.xml

```diff
@@ -524,312 +500,327 @@
+          don't provide context or meaning to IOAM-Data-Fields).</t>
+
+          <t>For a E2E Option-Type, here is the list of header fields that
+          participate (in that order) in the integrity protection of
+          IOAM-Data-Fields:
+            <list style="numbers">
+              <t>Namespace-ID</t>
+              <t>IOAM-E2E-Type</t>
+            </list>
+
+          Those are all the header fields defined for a E2E Option-Type.</t>
+
+          <t>A document defining a new IOAM Integrity Protected Option-Type that
+          intends to use the Integrity Protection Method defined in this
+          document MUST also specify a list of corresponding Option-Type header
+          fields that participate in the integrity protection of
+          IOAM-Data-Fields, as above. The same logic can be applied to select
+          header fields, by following this simple rule:
+            <list style="bullets">
+              <t>Only immutable IOAM Option-Type header fields that provide
+              context or meaning to IOAM-Data-Fields are considered, others are
+              excluded (e.g., mutable or reserved fields). For example, the
+              Namespace-ID, which is common to all IOAM Option-Types, MUST
+              always be included in the list. Once specified, the list MUST NOT
+              change for interoperability reasons.</t>
+            </list>
           </t>
+        </section>
+      </section>
 
-          <t>The decapsulating node MUST act as a Validator. The decapsulating
-          node MUST NOT generate a signature based on IOAM-Data-Fields it has
-          inserted, if any, and therefore MUST NOT update the Signature field.</t>
-        </list>
+      <section anchor="IPM-Transit" title="Transit node">
+        <t>For a transit node, the Integrity Check Value (ICV) is the result of
+        a GMAC operation over the received ICV field and immutable
+        IOAM-Data-Fields added by the transit node. In other words, with the
+        received Nonce field provided to GMAC, the transit node performs the
+        following operation:
+          <list style="bullets">
+            <t>AAD = (ICV field || node's immutable IOAM-Data-Fields)</t>
+            <t>ICV = GMAC(AAD)</t>
+          </list>
         </t>
-      </list>
-      </t>
 
-      <t>The method assumes that symmetric keys have been distributed to
-      the respective nodes as well as the Validator(s). The details of the
-      mechanisms used for key distribution are outside the scope of this
-      document.</t>
+        <t>The transit node updates the ICV field in the IOAM Integrity
+        Protection header. Reminder: it MUST always be a 16-octet ICV.</t>
+
+        <t>If the transit node does not add any IOAM-Data-Fields (e.g., it
+        only modifies mutable IOAM-Data-Fields or does nothing), then the
+        transit node MUST NOT update the ICV field in the IOAM Integrity
+        Protection header.</t>
```

Comment:
Is this really true? In case of pre-allocated trace, the fields would be *updated* and not added, but you still want the ICV field to be updated, wouldn't you?

Reply:
Indeed. Do you have a better wording in mind?

Reply:
How about:

"If the transit node does not add or change any IOAM-Data-Fields (e.g., it only modifies mutable IOAM-Data-Fields or does nothing), then the transit node MUST NOT update the ICV field in the IOAM Integrity Protection header."

Thinking about things again: Why do we even deal with this special case? There would be no harm for the node to update the ICV field - even if everything is unchanged. Plus you would know that the node was indeed visited (kind of "proof of transit" through a backdoor).

Reply:
> How about:
> 
> "If the transit node does not add or change any IOAM-Data-Fields (e.g., it only modifies mutable IOAM-Data-Fields or does nothing), then the transit node MUST NOT update the ICV field in the IOAM Integrity Protection header."

"or change" could be problematic (see below).

> Thinking about things again: Why do we even deal with this special case? There would be no harm for the node to update the ICV field - even if everything is unchanged. Plus you would know that the node was indeed visited (kind of "proof of transit" through a backdoor).

Think about an Option-Type with a mutable data field (currently, you only have the "Cumulative" field in POT type 0, but who knows what will come next). If a node modifies the value and updates the ICV, the integrity validation will always fail (there is no way to get previous values of the field). On the other hand, if everything is unchanged, it means the node cannot be identified in the data fields and so MUST NOT update the ICV, otherwise the integrity validation will always fail too (in addition to being a total waste of cpu cycles).

How about:

"If the transit node only modifies mutable IOAM-Data-Fields or does nothing, then the transit node MUST NOT update the ICV field in the IOAM Integrity Protection header."

In other words, is it better by removing the first part, i.e., "If the transit node does not add any IOAM-Data-Fields" ?

Reply:
Correct. The confusion comes from "mutable vs immutable fields" for Trace Option-Types. With Pre-allocated, it could indeed be considered mutable fields since it is first pre-allocated and then "modified" by the node. With Incremental, it can only be considered immutable fields. TBH, I don't know how to rephrase this to make it work :-D

Reply:
Given that the mechanism does not really work for fields like "cumulative", could we just limit ICV to trace and E2E option types?

Reply:
How about we remove it for now - and then ask the WG for either acknowledging - or proposing a solution which would work.

Reply:
Works for me. Note that in the first paragraph describing a transit node, we also have "[...] and **immutable** IOAM-Data-Fields **added** by the transit node". Same for the encapsulating node. We'll have to determine if we keep those as is (especially considering the pre-allocated trace case).